client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -365,7 +377,21 @@ private void ackUpTo(long ackLevel) {
             }
             state.releaseIfEmptyInflight();
         }
-        
+
+        private void checkAckLevels(long ackLevel, long previousAckLevel) {
+            if (previousAckLevel > ackLevel) {
+                throw new IllegalStateException("Bad ack from server - previousAckLevel = " +
```

Comment:
use `Preconditions.checkState` here and a few lines below. Also try to make use of string formatting (built into `checkState` (and siblings)) so that we don't need to do ugly string concat every time.

Reply:
fixed